lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,236 @@
+          Fog::Logger.warning("Template processed. Populated #{@files}")
+          @files
+        end
+
+        def files_basepath
+          min_lenght = @files.keys.map(&:length).min
+          candidates = @files.keys.map { |x| File.dirname(x[0..min_lenght]) }.to_set
+          return nil if candidates.size != 1
+          candidates.each do |x|
+            return x
+          end
+        end
+
+        def template_is_raw?(content)
+          # Return true if the file is an heat template, false otherwise.
+          htv = content.index("heat_template_version:")
+          !!(htv && htv < 5)
+        end
+
+        def template_is_url?(path)
+          # return true if it's an URI, false otherwise.
+          normalise_file_path_to_url(path)
+          return true
+        rescue ArgumentError, URI::InvalidURIError
+          return false
+        end
+
+        def file_outside_base_url(base_url, str_url)
+          ret = base_url && str_url.start_with?("file://") && !str_url.start_with?(base_url)
+          if ret
+            Fog::Logger.warning("Trying to reference a file outside #{base_url}: #{str_url}")
+          end
+          ret
+        end
+
+        def ignore_if(key, value)
+          # Return true if I should I process this this file.
+          #
+          # @param [String] An heat template key
+          #
+          return true if key != 'get_file' && key != 'type'
+
+          return true unless value.kind_of?(String)
+
+          return true if key == 'type' &&
+                         !value.end_with?('.yaml', '.template')
+
+          false
+        end
+
+        def recurse_if(value)
+          # Return true if I should inspect this yaml template branch.
+          value.kind_of?(Hash) || value.kind_of?(Array)
+        end
+
+        def url_join(prefix, suffix)
+          # Return string
+          if prefix
+            suffix = URI.join(prefix, suffix)
+            suffix = suffix.to_s.sub(%r{^file:\/+}, "file:///")
+          end
+          suffix
+        end
+
+        def base_url_for_url(url)
+          # Returns the string baseurl of the given url.
+          parsed = URI(url)
+          parsed_dir = File.dirname(parsed.path)
+          url_join(parsed, parsed_dir)
+        end
+
+        def normalise_file_path_to_url(path)
+          # Nothing to do on URIs
+          return path if URI(path).scheme
+
+          path = File.absolute_path(path)
+          url_join('file:/', path)
+        end
+
+        def get_content(uri_or_filename)
+          # Retrive the content of a local or remote file.
+          #
+          # @param A local or remote uri.
+          #
+          Fog::Logger.debug("Opening #{uri_or_filename}")
+          uri_or_filename = uri_or_filename.to_s if uri_or_filename.kind_of?(URI)
+          # throw exceptions enables stack creation to fail
+          #   with a suitable error.
+          #
+          # XXX Implement a retry here?
```

Comment:
This first implementation won't contain a retry. It could be added in the future.